### PR TITLE
`f16` and `f128` step 2: intrinsics

### PR DESCRIPTION
**Pull request by tgross35**
Continuation of https://github.com/rust-lang/rust/pull/121728, another portion of https://github.com/rust-lang/rust/pull/114607.

This PR adds `f16` and `f128` intrinsics, and hooks them up to both HIR and LLVM. This is all still unexposed to the frontend, which will probably be the next step. Also update itanium mangling per @rcvalle's in https://github.com/rust-lang/rust/pull/121728/files#r1506570300, and fix a typo from step 1.

Once these types are usable in code, I will add the codegen tests from #114607 (codegen is passing on that branch)

This does add more `unimplemented!`s to Clippy, but I still don't think we can do better until library support is added.

Tracking issue: https://github.com/rust-lang/rust/issues/116909

r? @compiler-errors
cc @Nilstrieb 
@rustbot label +T-compiler +F-f16_and_f128

### CODE REVIEW

File: src/tools/clippy/clippy_lints/src/approx_const.rs [hunk 1]
```diff
@@ -75,10 +75,12 @@ impl ApproxConstant {
     fn check_lit(&self, cx: &LateContext<'_>, lit: &LitKind, e: &Expr<'_>) {
         match *lit {
             LitKind::Float(s, LitFloatType::Suffixed(fty)) => match fty {
+                FloatTy::F16 => self.check_known_consts(cx, e, s, "f16"),
                 FloatTy::F32 => self.check_known_consts(cx, e, s, "f32"),
                 FloatTy::F64 => self.check_known_consts(cx, e, s, "f64"),
+                FloatTy::F128 => self.check_known_consts(cx, e, s, "f128"),
             },
-            LitKind::Float(s, LitFloatType::Unsuffixed) => self.check_known_consts(cx, e, s, "f{32, 64}"),
+            LitKind::Float(s, LitFloatType::Unsuffixed) => self.check_known_consts(cx, e, s, "f{16, 32, 64, 128}"),
```
Comment (compiler-errors) [inline src/tools/clippy/clippy_lints/src/approx_const.rs:83]: I think this should be a FIXME(f16_f128) with a "add these when f16/f128 are stable).

**Review by compiler-errors — CHANGES_REQUESTED** [review]
One nit, also needs CI to pass. r=me after.

File: src/tools/clippy/clippy_lints/src/approx_const.rs (hunk 1 above)
Comment (tgross35) [reply src/tools/clippy/clippy_lints/src/approx_const.rs]: Thanks, I think this was the failing CI as well. Fixed this

@rustbot ready

File: src/tools/clippy/clippy_lints/src/approx_const.rs [hunk 2]
```diff
@@ -75,9 +75,12 @@ impl ApproxConstant {
     fn check_lit(&self, cx: &LateContext<'_>, lit: &LitKind, e: &Expr<'_>) {
         match *lit {
             LitKind::Float(s, LitFloatType::Suffixed(fty)) => match fty {
+                FloatTy::F16 => self.check_known_consts(cx, e, s, "f16"),
                 FloatTy::F32 => self.check_known_consts(cx, e, s, "f32"),
                 FloatTy::F64 => self.check_known_consts(cx, e, s, "f64"),
+                FloatTy::F128 => self.check_known_consts(cx, e, s, "f128"),
             },
+            // FIXME(f16_f128): add these when f16/f128 are available in library
```
Comment (compiler-errors) [inline src/tools/clippy/clippy_lints/src/approx_const.rs:83]: Not available in the standard library, but stable. We shouldn't be leaking unstable features into clippy outputs.

**Review by compiler-errors — APPROVED** [review]

File: src/tools/clippy/clippy_lints/src/approx_const.rs (hunk 2 above)
Comment (tgross35) [reply src/tools/clippy/clippy_lints/src/approx_const.rs:83]: Ah, makes sense. I'll get this in the next round

File: compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs [hunk 3]
```diff
@@ -464,16 +464,17 @@ fn encode_ty<'tcx>(
...
         //
         // (See https://rust-lang.github.io/unsafe-code-guidelines/layout/scalars.html#fixed-width-floating-point-types.)
         ty::Float(float_ty) => {
-            typeid.push(match float_ty {
-                FloatTy::F16 => unimplemented!("f16_f128"),
-                FloatTy::F32 => 'f',
-                FloatTy::F64 => 'd',
-                FloatTy::F128 => unimplemented!("f16_f128"),
+            typeid.push_str(match float_ty {
+                FloatTy::F16 => "Dh",
```
[the remark this answers is not in the document]
Comment (tgross35) [reply compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs:474]: Sure, I'll get this in the next round. Thanks for taking a look!

**Comment by compiler-errors** [comment]
Actually just fix both of these issues now, there's no rush to get this PR merged.

@bors r-

**Comment by tgross35** [comment]
Easy enough, fixed both things

**Review by compiler-errors — APPROVED** [review]

**Pull request merged**